daisi/src/minhton-ns3/scenario_steps.h

```diff
@@ -0,0 +1,207 @@
+// Copyright 2023 The SOLA authors
+//
+// This file is part of DAISI.
+//
+// DAISI is free software: you can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation; version 2.
+//
+// DAISI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+// the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+// Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along with DAISI. If not, see
+// <https://www.gnu.org/licenses/>.
+//
+// SPDX-License-Identifier: GPL-2.0-only
+
+#ifndef DAISI_MINHTON_NS3_MINHTON_SCENARIO_STEPS_H_
+#define DAISI_MINHTON_NS3_MINHTON_SCENARIO_STEPS_H_
+
+#include <cstdint>
+#include <optional>
+#include <string>
+
+#include "manager/scenariofile_component.h"
+
+#define TYPE_NAME(name) \
+  static std::string typeName() { return name; }
+
+namespace daisi::minhton_ns3 {
+
+struct JoinOne {
+  TYPE_NAME("join-one")
+
+  std::optional<uint32_t> delay;
+  std::optional<uint32_t> level;
+  std::optional<uint32_t> number;
+  std::optional<uint32_t> index;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(level);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(index);
+  }
+  static std::string getTypeName() { return "join-one"; }
+};
+
+struct JoinMany {
+  TYPE_NAME("join-many")
+
+  std::optional<uint32_t> delay;
+  uint32_t number;
+  std::string mode;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(mode);
+  }
+};
+
+struct LeaveOne {
+  TYPE_NAME("leave-one")
+
+  std::optional<uint32_t> delay;
+  std::optional<uint32_t> level;
+  std::optional<uint32_t> number;
+  std::optional<uint32_t> index;
+
+  void parse(YAML::Node node) {
+    SERIALIZE_VAR(delay);
+    SERIALIZE_VAR(level);
+    SERIALIZE_VAR(number);
+    SERIALIZE_VAR(index);
+  }
+};
+
+struct LeaveMany {
+  TYPE_NAME("leave-many");
+  std::optional<uint32_t> delay;
```

Comment:
One free line in between please to keep it consistent and clean ;)

```suggestion
  TYPE_NAME("leave-many");
  
  std::optional<uint32_t> delay;
```